### ISSUE TITLE
Fix Chart Load Performance

### ISSUE THREAD

**Issue by dbpiper** [dbpiper/Euclid#91]
# User Story

As a user I would like the page to load quickly so that I can use it and not abandon the site.

## Proposed changes
The site is currently incredibly slow!!! It takes about 10-15 seconds for the Chart to load 3 year data...
This is obviously very bad for UX, and getting people (including myself) to use the site! In fact, [40% of people will abandon a site that takes more than 3 seconds to load.](https://neilpatel.com/blog/loading-time/) In other words, the site is about 5 times slower than it should be. Essentially, it is so slow that _no one_ would use it, even if it were fully functional...

* **What is the current behavior?** (You can also link to an open issue here)
The site is **way** too slow, specifically the chart load times and the drop-down load times.

* **What is the new behavior (if this is a feature change)?**
The site loads quickly.

## Types of changes

What types of changes does your code introduce to Euclid?

* [x] Bug-fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Steps to Reproduce

1. Load the page.
2. Observe that it takes about 15 seconds to load the chart data.
3. This can also be seen in the end-to-end test times.

**Comment by dbpiper** [comment]
Some initial observations from the verify script.

The Storybook chart integration tests take about 5 seconds each.

The e2e tests each take about 14-20 seconds each. 

This seems to indicate that _both_ the rendering of the data and its delivery/lookup are the problems. Which means that both should be investigated to solve this issue.

**Comment by dbpiper** [comment]
![chart integration test times.png](https://images.zenhubusercontent.com/58e14e7251aa2c30577a2552/5d38de44-35e7-4eea-a72c-975d24625a62)![integration test header times.png](https://images.zenhubusercontent.com/58e14e7251aa2c30577a2552/8976acb3-06e9-4961-80cb-c786b720f8bc)![e2e test selector times.png](https://images.zenhubusercontent.com/58e14e7251aa2c30577a2552/29b76e60-1751-4945-9a9f-1e15a36762e4)

**Comment by dbpiper** [comment]
Attached some screenshots demonstrating these test times.

**Comment by dbpiper** [comment]
![chart integration test (no animation).png](https://images.zenhubusercontent.com/58e14e7251aa2c30577a2552/d89c92a9-143e-459c-a808-2c812a294f56)

I ran an integration test and captured the times, for the chart with no animation. This is the raw, drawing/data loading time; without any network requests being made. In other words, this is the floor to the loading time of the chart. 

As expected, the animation adds 1.5 seconds to the testing times (since the default Recharts animation time is 1.5 seconds). 

This is quite disturbing since the 5 years/All Time data is about 3.1 seconds with no animation and no network requests. It looks like the 3 Years data is being cached, since it is the default view. 

In other words, the "3 Years works" test isn't actually very representative of the true 3 years performance. Rather the "the default view works" test is.

In other words, if the goal is <3 seconds, everything but 5 Years/All time can potentially work. The 6 Months/YTD - 3 Years is all well under 1.5 seconds, without animation or network requests, which is good.

However, the question is: is it possible for me to lower the actual page performance to get anywhere close to these numbers. In other words, can I somehow make the real site get close to the Storybook speeds.

**Comment by dbpiper** [comment]
After hours on Friday I did a bit of profiling of the server, specifically of the resolvers (getting the data from the database). I discovered that these are **not** the problem and they take less than a second each, which is good.

So on Monday I will be heavily profiling the client, as I am pretty sure that it is the problem.

Specifically:

- how long it takes to send the query and get the response (total transit time)
- how long it takes to take the data that was fetched from the server and display it on the chart

**Comment by dbpiper** [comment]
I have done some profiling on the client, per the previous comment.

I have discovered (not surprisingly) that it takes almost no time to fetch the data for the tickers and about 15 seconds to fetch the chart data. Now the odd thing is the fact that the tickers take about 15 seconds to load in the e2e tests.![client profiling information.png](https://images.zenhubusercontent.com/58e14e7251aa2c30577a2552/6fec2fdd-c434-4ffe-8b52-2dde8f0c651b)

**Comment by dbpiper** [comment]
After a bit more investigation, I have discovered that the stock data is actually fetched correctly in a reasonable amount of time (< 2 seconds), however for some reason the client keeps spamming the sever for the data over and over again. 

This can be seen in this screenshot.

![multiple stock loading... .png](https://images.zenhubusercontent.com/58e14e7251aa2c30577a2552/7518238f-1763-4691-a9ed-c19e0b1d8e84)

**Comment by dbpiper** [comment]
I just confirmed that the client is only hitting the server once for a very short amount of time. Which means that it is **definitely** a problem with the client itself. Furthermore, it can't be some sort of very long network request since the server only gets the initial (1 second) request.

Very mysterious. However, it is clear that there is some weird bug in the chart loading logic (as the tickers load just fine by themselves if I comment out the ChartGroup).

![only hitting the server once... .png](https://images.zenhubusercontent.com/58e14e7251aa2c30577a2552/4585e68a-c150-4560-b4f2-d613f048c92c)

**Comment by dbpiper** [comment]
I believe that I have figured out the problem, there was an O(n * lg(n)) algorithm (a sort of the stocks) happening a ton of times in the renderer. Specifically, it was doing this operation which was not needed (as the result can easily be cached/memoized) at least 10+ times (for each of the visible X Ticks).

Once I fixed this, everything is fast as expected!!

I still have a bit of cleanup work to do, but this should solve the problem once it is correctly implemented.

**Comment by dbpiper** [comment]
I went ahead and memoized all of the utility functions inside of `Home/util` with [memoize-one](https://github.com/alexreardon/memoize-one) That is all of the functions in `date-formatting` and `stock-formatting`.

This completely resolves the slowness issue which in turn, fixes this issue!
